sql/api/src/main/scala/org/apache/spark/sql/types/GeometryType.scala

```diff
@@ -0,0 +1,209 @@
+   */
+  override def equals(obj: Any): Boolean = {
+    obj match {
+      case g: GeometryType =>
+        // Iff two GeometryTypes have the same SRID, they are considered equal.
+        g.srid == srid
+      case _ =>
+        // In all other cases, the two types are considered not equal.
+        false
+    }
+  }
+
+  /**
+   * The hash code of the GeometryType is derived from its SRID value.
+   */
+  override def hashCode(): Int = srid.hashCode
+
+  /**
+   * The GeometryType can only accept another type if the other type is also a GeometryType,
+   * and the SRID values are compatible (see `acceptsGeometryType` below for more details).
+   */
+  override private[sql] def acceptsType(other: DataType): Boolean = {
+    other match {
+      case gt: GeometryType =>
+        // For GeometryType, we need to check the SRID values.
+        acceptsGeometryType(gt)
+      case _ =>
+        // In all other cases, the two types are considered different.
+        false
+    }
+  }
+
+  /**
+   * The GeometryType with a mixed SRID can accept any other GeometryType, i.e. either a fixed SRID
+   * GeometryType or another mixed SRID GeometryType. Conversely, a GeometryType with a fixed SRID
+   * can only accept another GeometryType with the same fixed SRID value, and not a mixed SRID.
+   */
+  def acceptsGeometryType(gt: GeometryType): Boolean = {
+    // If the SRID is mixed, we can accept any other GeometryType.
+    // If the SRID is not mixed, we can only accept the same SRID.
+    isMixedSrid || gt.srid == srid
+  }
+
+  /**
+   * Converts a CRS string to its corresponding SRID integer value.
+   */
+  private[types] def toSrid(crs: String): Int = {
+    // The special value "SRID:ANY" is used to represent mixed SRID values.
+    if (crs.equalsIgnoreCase(GeometryType.GEOMETRY_MIXED_CRS)) {
+      return GeometryType.GEOMETRY_MIXED_SRID
+    }
+    // For all other CRS values, we need to look up the corresponding SRID.
+    val srid = CrsMappings.get().getSrid(crs)
+    if (srid == null) {
+      // If the CRS value is not recognized, we throw an exception.
+      throw new SparkIllegalArgumentException(
+        errorClass = "ST_INVALID_CRS_VALUE",
+        messageParameters = Map("crs" -> crs)
+      )
+    }
+    srid
+  }
+}
+
+@Experimental
+object GeometryType extends AbstractDataType {
+
+  /**
+   * Mixed SRID value and the corresponding CRS value for GeometryType.
+   * This value is completely internal, and should not be exposed to users.
+   * It means we can have different SRIDs per row.
+   */
+  final val GEOMETRY_MIXED_SRID = -1
+  final val GEOMETRY_MIXED_CRS = "SRID:ANY"
+
+  /**
+   * The default coordinate reference system (CRS) value used for geometries,
+   * as specified by the Parquet, Delta, and Iceberg specifications.
+   * If crs is omitted, it should always default to this.
+   */
+  final val DEFAULT_STORAGE_CRS = "OGC:CRS84"
```

Comment:
nit: should be GEOMETRY_DEFAULT_CRS?  (to match Geography).  I think be nice to extract to common